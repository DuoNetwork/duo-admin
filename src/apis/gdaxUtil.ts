import sqlUtil from '../sqlUtil';
import util from '../util';
import * as CST from '../constants';
import { Trade } from '../types';

const INTERVAL_SECS = 2;

export class GdaxUtil {

	parseTrade(trade: {[key: string]: string}): Trade {
		return {
			source: CST.EXCHANGE_GDAX,
			id: trade.trade_id,
			price: Number(trade.price),
			amount: Math.abs(Number(trade.size)),
			timestamp: new Date(trade.time).valueOf()
		};
	}

	async fetchETHTradesByRestfulAPI() {
		const data = await util.get('https://api.gdax.com:443/products/ETH-USD/trades');
		const parsedData: Array<{ [key: string]: string }> = JSON.parse(data);

		parsedData.forEach(item => {
			// console.log(item);
			sqlUtil.insertSourceData(
				this.parseTrade(item)
			);
		});
	}

	startFetching() {
		setInterval(() => this.fetchETHTradesByRestfulAPI(), INTERVAL_SECS * 1000);
	}
}

const gdaxUtil = new GdaxUtil();
export default gdaxUtil;
