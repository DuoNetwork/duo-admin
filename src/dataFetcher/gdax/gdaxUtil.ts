import MysqlUtil from '../../utils/MysqlUtil';
import * as CST from '../../constant';

const INTERVAL_SECS = 2;

const EXCHANGE_NAME = CST.EXCHANGE_GDAX;
const DB_HOST = CST.DB_HOST;
const DB_USER = CST.DB_USER;
const DB_PASSWORD = CST.DB_PASSWORD;
const DB_PRICEFEED = CST.DB_PRICEFEED;
const DB_TABLE_TRADE = CST.DB_TABLE_TRADE;

export class CoinbaseGDAXTradeFeedUtil {
	mysqlUtil: MysqlUtil;

	constructor() {
		this.mysqlUtil = new MysqlUtil(EXCHANGE_NAME, DB_HOST, DB_USER, DB_PASSWORD, DB_PRICEFEED, DB_TABLE_TRADE);
	}

	initDB() {
		console.log('Init the DB');
		this.mysqlUtil.initDB();
	}

	fetchETHTradesByRestfulAPI() {
		var https = require('https');

		var options = {
			host: 'api.gdax.com',
			path: '/products/ETH-USD/trades',
			port: '443',
			//This is the only line that is new. `headers` is an object with the headers to request
			headers: { 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10; rv:33.0) Gecko/20100101 Firefox/33.0' }
		};

		var callbackFunc = response => {
			var responseStr = '';
			response.on('data', chunk => {
				responseStr += chunk;
			});

			response.on('end', () => {
				var parsedJson = JSON.parse(responseStr);

				let dbConn = coinbaseGDAXTradeFeedUtil.mysqlUtil.dbConn;
				if (dbConn == undefined) {
					coinbaseGDAXTradeFeedUtil.initDB();
				}

				parsedJson.forEach(item => {
					coinbaseGDAXTradeFeedUtil.mysqlUtil.insertDataIntoMysql(EXCHANGE_NAME, item.trade_id, item.price, item.size, item.side, new Date(item.time).valueOf() +"");
				});
			});
		};

		var req = https.request(options, callbackFunc);
		req.end();
	}

	startFetching() {
		setInterval(coinbaseGDAXTradeFeedUtil.fetchETHTradesByRestfulAPI, INTERVAL_SECS * 1000);
	}
}

const coinbaseGDAXTradeFeedUtil = new CoinbaseGDAXTradeFeedUtil();
export default coinbaseGDAXTradeFeedUtil;
