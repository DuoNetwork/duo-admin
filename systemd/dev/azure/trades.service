[Unit]
Description=DUO Azure Trades Service
Documentation=https://github.com/FinBook/duo-admin
After=network.target
 
[Service]
Environment="PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin"
WorkingDirectory=/home/ubuntu/duo-admin
Type=simple
User=ubuntu
ExecStart=/usr/bin/node /home/ubuntu/duo-admin/dist/index trades assets=ETH,USD server azure
Restart=always
 
[Install]
WantedBy=multi-user.target