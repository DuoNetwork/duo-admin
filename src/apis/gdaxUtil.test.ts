import trades from '../samples/gdax.json';
import gdaxUtil from './gdaxUtil';

test('parseTrade', () => {
	trades.forEach(async trade => {
		const parsedTrade = gdaxUtil.parseTrade(trade);
		expect(parsedTrade).toMatchSnapshot();
	});
});
