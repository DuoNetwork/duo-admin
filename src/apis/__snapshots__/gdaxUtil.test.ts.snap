// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTrade 1`] = `
Object {
  "amount": 0.1198,
  "id": "32872596",
  "price": 630.17,
  "source": "GDAX",
  "timestamp": 1524739231228,
}
`;

exports[`parseTrade 2`] = `
Object {
  "amount": 0.983,
  "id": "32872595",
  "price": 630.11,
  "source": "GDAX",
  "timestamp": 1524739231228,
}
`;

exports[`parseTrade 3`] = `
Object {
  "amount": 0.2,
  "id": "32872594",
  "price": 630.11,
  "source": "GDAX",
  "timestamp": 1524739231228,
}
`;
